### PR TITLE
Support STM32H7 flash storage

### PR DESCRIPTION
**Pull request by tridge**
This adds support for AP_FlashStorage for STM32H7. This is needed for a new H743 based Matek board
The H7 is tricky as:
 - you must write 32 bytes at a time
 - you can't write to a block that isn't all 0xFF without risking ECC corruption
 - sectors are 128k in size

This PR abstracts AP_FlashStorage to allow for the 3 different types of storage more cleanly

This has been tested on:
- f303-GPS for F3 support
- MatekF405-Wing for F4 support
- MatekH743 for H7 support

### CODE REVIEW

**Comment by peterbarker** [comment]
On Sat, 11 Apr 2020, Andrew Tridgell wrote:

>  *  sectors are 128k in size

My understanding is that the erase time *is* deterministic.  But is it 
close to our watchdog-poking-interval?

Peter

File: libraries/AP_FlashStorage/AP_FlashStorage.cpp [hunk 1]
```diff
@@ -286,6 +308,9 @@ bool AP_FlashStorage::load_sector(uint8_t sector)
             // invalid state
             return false;
         }
+#if AP_FLASHSTORAGE_TYPE == AP_FLASHSTORAGE_TYPE_H7
+        ofs = (ofs + 31U) & ~31U;
```
Comment (magicrub) [inline libraries/AP_FlashStorage/AP_FlashStorage.cpp:312]: & -31 ???

This, at the very least, needs a comment

File: libraries/AP_FlashStorage/AP_FlashStorage.h [hunk 2]
```diff
@@ -39,24 +39,47 @@
...
+#define AP_FLASHSTORAGE_TYPE AP_FLASHSTORAGE_TYPE_F1
+#elif defined(STM32H7)
+/*
+  STM32H7 can only write in 32 byte chunks, and must only write when all bits are 1
+ */
+#define AP_FLASHSTORAGE_TYPE AP_FLASHSTORAGE_TYPE_H7
 #else
-#define AP_FLASHSTORAGE_MULTI_WRITE 1
+/*
+  STM32HF4 and STM32HF7 can update bits from 1 to 0
```
Comment (magicrub) [inline libraries/AP_FlashStorage/AP_FlashStorage.h:63]: Comment is wrong. Says STM32HF

File: libraries/AP_FlashStorage/AP_FlashStorage.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_FlashStorage/AP_FlashStorage.cpp:315]: comment added

File: libraries/AP_FlashStorage/AP_FlashStorage.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_FlashStorage/AP_FlashStorage.h]: fixed

**Comment by tridge** [comment]
> But is it close to our watchdog-poking-interval?

it is under the watchdog interval even with 128k sectors. I've added an expect delay call around the erase

File: libraries/AP_FlashStorage/AP_FlashStorage.cpp [hunk 3]
```diff
@@ -197,27 +201,45 @@ bool AP_FlashStorage::write(uint16_t offset, uint16_t length)
...
-#if AP_FLASHSTORAGE_MULTI_WRITE
-        if (!flash_write(current_sector, write_offset, (uint8_t*)&header, sizeof(header))) {
+
+        struct PACKED {
+            struct block_header header;
+            uint8_t data[max_write];
+        } blk;
+
+        blk.header.state = BLOCK_STATE_WRITING;
+        blk.header.block_num = offset / block_size;
```
Comment (rmackay9) [inline libraries/AP_FlashStorage/AP_FlashStorage.cpp:211]: no chance of block_size being zero of course?

File: libraries/AP_FlashStorage/AP_FlashStorage.cpp [hunk 4]
```diff
@@ -227,7 +249,9 @@ bool AP_FlashStorage::write(uint16_t offset, uint16_t length)
         offset += n2;
         length -= n2;
     }
-    
+
+    //debug("write_offset %u\n", write_offset);
```
Comment (rmackay9) [inline libraries/AP_FlashStorage/AP_FlashStorage.cpp:253]: little debug here..

**Pull request merged**